### PR TITLE
[Forwardport] Ensure integer values are not quoted as strings

### PR DESCRIPTION
**Pull request by gelanivishal**
### Original Pull Request
[#18287](https://github.com/magento/magento2/pull/18287)

### Description

In cases where indexers does a filtering in the style of `WHERE entity_id IN (...)` there is a performance penalty if IDs are passed as strings, rather than integer values. This is especially visible on larger data sets. This pull request should cover some common occurrences of this.

### Fixed Issues (if relevant)

1. n/a

### Manual testing scenarios

1. n/a

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @gelanivishal. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me $VERSION instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Review by DanielRuf — CHANGES_REQUESTED** [review]
The lines are now longer than 80 chars and readability is partly worse now.

Also int typecasting is faster than intval.

File: lib/internal/Magento/Framework/Mview/View.php [hunk 1]
```diff
@@ -284,8 +284,8 @@ public function update()
 
                 for ($vsFrom = $lastVersionId; $vsFrom < $currentVersionId; $vsFrom += $versionBatchSize) {
                     // Don't go past the current version for atomicy.
-                    $versionTo = min($currentVersionId, $vsFrom + $versionBatchSize);
-                    $ids = array_map('intval', $this->getChangelog()->getList($vsFrom, $versionTo));
+                    $versionTo = min($currentVersionId, $versionFrom + $versionBatchSize);
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Mview/View.php:287]: Looks like `$versionFrom` is not defined there. Please fix it.

File: app/code/Magento/Catalog/Model/ProductCategoryList.php [hunk 2]
```diff
@@ -80,10 +80,7 @@ public function getCategoryIds($productId)
                 Select::SQL_UNION_ALL
             );
 
-            $this->categoryIdList[$productId] = array_map(
-                'intval',
-                $this->productResource->getConnection()->fetchCol($unionSelect)
-            );
+            $this->categoryIdList[$productId] = array_map('intval', $this->productResource->getConnection()->fetchCol($unionSelect));
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/ProductCategoryList.php]: Previously code was the same as in 2.2-develop, so it's better to keep it as it was before. Please revert this change

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/Mview/View.php (hunk 1 above)
Comment (gelanivishal) [reply lib/internal/Magento/Framework/Mview/View.php:287]: @ihor-sviziev Done.

File: app/code/Magento/Catalog/Model/ProductCategoryList.php (hunk 2 above)
Comment (gelanivishal) [reply app/code/Magento/Catalog/Model/ProductCategoryList.php]: @ihor-sviziev Done.

**Comment by gelanivishal** [comment]
@ihor-sviziev I have performed your requested changes.

**Comment by DanielRuf** [comment]
Reverted? Does not seem so: https://github.com/magento/magento2/pull/18961/files

**Comment by gelanivishal** [comment]
@DanielRuf Check now.

File: app/code/Magento/Catalog/Model/ResourceModel/Product/Collection.php [hunk 3]
```diff
@@ -2021,11 +2021,7 @@ protected function _applyProductLimitations()
             $this->getConnection()->quoteInto('cat_index.store_id=?', $filters['store_id'], 'int'),
         ];
         if (isset($filters['visibility']) && !isset($filters['store_table'])) {
-            $conditions[] = $this->getConnection()->quoteInto(
```
Comment (DanielRuf) [inline app/code/Magento/Catalog/Model/ResourceModel/Product/Collection.php:2024]: Still not reverted here, please use the more readable version with multiple lines

**Comment by DanielRuf** [comment]
In sum, there is currently no `Ensure integer values are not quoted as strings`. I highly advise not to hijack PRs of others and doing more changes. This here is not a forward port imho.

**Comment by orlangur** [comment]
@gelanivishal please fix this PR branch using force push so that it contains only two commits exactly as https://github.com/magento/magento2/pull/18287/commits preserving original authors. Otherwise we will have to reject such forwardport.
